docs(pathops): extended readme (#55)
canonical/charmtech-charmlibs: pathops/README.md

```

Check out the reference docs [here](https://canonical-charmlibs.readthedocs-hosted.com/reference/pathops/).

# Getting started

> To get started, consider creating an attribute with the root directory for the local or remote paths you'll be working with.

Author: Thanks, I'll tweak your suggestion to reflect that this is not the only way.
```suggestion
To get started, you can create an attribute `self.root` with the root directory for the local or remote paths you'll be working with.
```

```py
import ops
from charmlibs import pathops

class KubernetesCharm(ops.CharmBase):
    def __init__(self, framework: ops.Framework):
        super().__init__(framework)
        container = self.unit.get_container('container-name')
        self.root = pathops.ContainerPath('/', container=container)

# or

class MachineCharm(ops.CharmBase):
    def __init__(self, framework: ops.Framework):
        super().__init__(framework)
        self.root = pathops.LocalPath('/')
```

> You can then write code that could be used by either charm, using the `pathops.PathProtocol` type, which describes the operations supported by both local and container paths.

Author: I think I'll stick with what I have for this one, as it hopefully makes the reason to use `pathops.PathProtocol` over `LocalPath | ContainerPath` more clear (though I don't think drilling into this in detail is needed here).

```py
CONFIG_FILE_LOCATION = '/foo/bar'
CONFIG_FILE_CONTENTS = '...'

def write_config_file(root: pathops.PathProtocol) -> None:
    path = root / CONFIG_FILE_LOCATION

>     path.write_text(CONFIG_FILE_CONTENTS)

Author: I think I'll keep this one as the somewhat more concise original version. I guess the helper function could just be in-lined into `_on_config_changed`, but I figure maybe you'd do this on pebble-ready or some other startup event (the config file doesn't necessarily reflect Juju user exposed config, right?).

>     path.write_text(CONFIG_FILE_CONTENTS)

Reviewer: Makes sense. A dedicated how-to guide could cover a more specific/detailed scenario

>     path.write_text(CONFIG_FILE_CONTENTS)

Reviewer: Or, conversely, we find the relevant places in the Ops docs and promote pathops there :wink:

>     path.write_text(CONFIG_FILE_CONTENTS)

Author: Yeah that would be nice!

        super().__init__(framework)
        self.charm_root = pathops.LocalPath('/')
        container = self.unit.get_container('container-name')
        self.workload_root = pathops.ContainerPath('/', container=container)
        ...

    def copy_config(self):
        source = self.charm_root / 'foo/bar'
        dest = self.workload_root / 'foo/bar/baz'
        dest.parent.mkdir(parents=True, exist_ok=True)
        dest.write_bytes(source.read_bytes())
```

`pathops` also provides a helper function for ensuring that a file is present with some specified content, which returns whether changes were made.

```py
import ops
from charmlibs import pathops

class KubernetesCharm(ops.CharmBase):
    ...

    def copy_config(self):
        changed = pathops.ensure_contents(
            path=self.workload_root / 'foo/bar/baz',
            source=(self.charm_root / 'foo/bar').read_bytes(),
        )
        if changed:
            # e.g. make the workload reload its configuration
            ...
```

`pathops.PathProtocol` provides a subset of the `pathlib.Path` API -- relative paths, hardlink and symlink manipulation, and `open` are unsupported.

> A separate `chmod` method is also unsupported, as Pebble does not currently support this. `mkdir`, `write_bytes` and `write_text` instead provide the arguments `mode`, `user`, and `group` to set directory or file permissions and ownership. `ensure_contents` also supports these arguments.

Reviewer: Some small suggestions for readability:

```suggestion
`pathops` doesn't provide a separate `chmod` method, as Pebble doesn't currently support this. Instead, `mkdir`, `write_bytes`, and `write_text` have arguments `mode`, `user`, and `group` to set directory or file permissions and ownership. `ensure_contents` also has these arguments.
```
